### PR TITLE
Fc 475 get all apprenticeship favourites api

### PR DESCRIPTION
**Pull request by desderry**

### CODE REVIEW

File: src/api/EmployerFavouritesApi/Startup.cs [hunk 1]
```diff
@@ -27,12 +33,21 @@ public Startup(IConfiguration configuration, IHostingEnvironment env)
...
+            services.AddMediatR(typeof(SaveApprenticeshipFavouriteCommand).Assembly);
+
             services.AddMvc(options => {
                 if (!Environment.IsDevelopment())
                 {
                     options.Filters.Add(new AuthorizeFilter("default"));
                 }
             }).SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
+
+            services.AddSingleton<IAccountApiConfiguration>(x => Configuration.GetSection("AccountApiConfiguration").Get<AccountApiConfiguration>());
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Startup.cs:48]: Is the AccountApi configuration required?

File: src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            try
+            {
+
+                var apprenticeships = await _mediator.Send(new GetApprenticeshipFavouritesRequest() {EmployerAccountID = employerAccountId});
+
+                return Ok(apprenticeships);
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs]: I think this should be logging to the _logger instance.

File: src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                return BadRequest();
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs:59]: Is this a correct status code for all exceptions? This implies the error is always with caller and how they've made the request. 
We throw an `ArgumentException` in the handler if the accountId is invalid. Should this return a 400 and anything else return a 500?

File: src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        private readonly ILogger<ApprenticeshipsController> _logger;
+        private readonly IMediator _mediator;
+
+        public ApprenticeshipsController(ILogger<ApprenticeshipsController> logger, IMediator mediator)
+        {
+            _logger = logger;
+            _mediator = mediator;
+        }
+
+        // GET api/values
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs]: This comment is incorrect.

File: src/das-employer-favourites.ApplicationServices/Commands/GetApprenticeshipFavouritesRequest.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+using DfE.EmployerFavourites.ApplicationServices.Domain;
+using MediatR;
+
+namespace DfE.EmployerFavourites.ApplicationServices.Commands
+{
+    public class GetApprenticeshipFavouritesRequest : IRequest<ApprenticeshipFavourites>
```
Comment (leewadhams) [inline src/das-employer-favourites.ApplicationServices/Commands/GetApprenticeshipFavouritesRequest.cs]: I'd be tempted to separate the Commands and Queries into different folders just for some separation.

File: src/das-employer-favourites.ApplicationServices/Commands/GetApprenticeshipFavouriteRequestHandler.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using DfE.EmployerFavourites.ApplicationServices.Domain;
+using MediatR;
+using Microsoft.Extensions.Logging;
+using SFA.DAS.EAS.Account.Api.Client;
```
Comment (leewadhams) [inline src/das-employer-favourites.ApplicationServices/Commands/GetApprenticeshipFavouriteRequestHandler.cs]: Unused using

File: src/das-employer-favourites.ApplicationServices/Commands/GetApprenticeshipFavouriteRequestHandler.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        private readonly ILogger<GetApprenticeshipFavouriteRequestHandler> _logger;
+        private readonly IFavouritesRepository _repository;
+        private readonly IEmployerAccountRepository _employerAccountRepository;
+
+        public GetApprenticeshipFavouriteRequestHandler(
+            ILogger<GetApprenticeshipFavouriteRequestHandler> logger,
+            IFavouritesRepository repository, IEmployerAccountRepository employerAccountRepository)
+        {
+            _logger = logger;
```
Comment (leewadhams) [inline src/das-employer-favourites.ApplicationServices/Commands/GetApprenticeshipFavouriteRequestHandler.cs]: *_logger* isn't used currently. Should we be logging what accountId has had a request made somewhere?

File: src/das-employer-favourites.ApplicationServices/Commands/GetApprenticeshipFavouriteRequestHandler.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+using Microsoft.Extensions.Logging;
+using SFA.DAS.EAS.Account.Api.Client;
+
+namespace DfE.EmployerFavourites.ApplicationServices.Commands
+{
+    public class GetApprenticeshipFavouriteRequestHandler : IRequestHandler<GetApprenticeshipFavouritesRequest,ApprenticeshipFavourites>
+    {
+        private readonly ILogger<GetApprenticeshipFavouriteRequestHandler> _logger;
+        private readonly IFavouritesRepository _repository;
+        private readonly IEmployerAccountRepository _employerAccountRepository;
```
Comment (leewadhams) [inline src/das-employer-favourites.ApplicationServices/Commands/GetApprenticeshipFavouriteRequestHandler.cs]: Isn't used

File: src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<string>>> Get(string employerAccountId)
+        {
+            try
+            {
+
+                var apprenticeships = await _mediator.Send(new GetApprenticeshipFavouritesRequest() {EmployerAccountID = employerAccountId});
+
+                return Ok(apprenticeships);
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs]: I thought we were going to return the name of the apprenticeship too?

File: src/das-employer-favourites.ApplicationServices/Configuration/ConnectionStrings.cs [hunk 10]
```diff
@@ -0,0 +1,8 @@
+namespace DfE.EmployerFavourites.ApplicationServices.Configuration
```
Comment (leewadhams) [inline src/das-employer-favourites.ApplicationServices/Configuration/ConnectionStrings.cs:1]: If we're following the onion architecture layering this should live in Infrastructure

**Review by leewadhams — CHANGES_REQUESTED** [review]
I've still got reservations about the need of the ApplicationServices project. 
As part of a story we need to make the Web call the api to create the favourites. This will mean that every class in ApplicationServices is only used by 1 of the 2 apps. Lets discuss 👯‍♂ 

Is there a client package being created?

File: src/api/EmployerFavouritesApi/Startup.cs (hunk 1 above)
Comment (desderry) [reply src/api/EmployerFavouritesApi/Startup.cs:48]: this is for the MA employer account api so i believe it is

File: src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs [hunk 11]
```diff
@@ -23,22 +25,46 @@ public ApprenticeshipsController(ILogger<ApprenticeshipsController> logger, IMed
...
-            catch (Exception e)
+
+            catch (ArgumentException e)
             {
                 Console.WriteLine(e);
                 return BadRequest();
             }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs]: This is still using the Console to log the error

File: src/api/EmployerFavouritesApi/nlog.config [hunk 12]
```diff
@@ -13,9 +13,9 @@
   <!-- the targets to write to -->
   <targets>
 
-    <target xsi:type="Redis" name="redis" connectionStringName="Redis" environmentKeyName="EnvironmentName" appName="das-employer-favourites-api" includeAllProperties="true" layout="${message}">
+    <target xsi:type="Redis" name="redis" connectionStringTitle="Redis" environmentKeyTitle="EnvironmentTitle" appTitle="das-employer-favourites-api" includeAllProperties="true" layout="${message}">
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/nlog.config]: These renames aren't correct are they?

File: src/das-employer-favourites.ApplicationServices/Domain/IFatRepository.cs [hunk 13]
```diff
@@ -0,0 +1,9 @@
+using System.Threading.Tasks;
+
+namespace DfE.EmployerFavourites.ApplicationServices.Domain
+{
+    public interface IFatRepository
```
Comment (leewadhams) [inline src/das-employer-favourites.ApplicationServices/Domain/IFatRepository.cs]: I think this should live in Infrastructure. I don't think it's specifically a domain service. On it's own. I think the Implementation of the favourites domain service just uses a number of sources.

File: src/das-employer-favourites.ApplicationServices/Infrastructure/FatApiRepository.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+            {
+                return _standardApiClient.Get(apprenticeshipId).Title;
+            }
+            else
+            {
+                return _frameworkApiClient.Get(apprenticeshipId).Title;
+            }
+        }
+
+        private bool isStandard(string apprenticeshipId)
```
Comment (leewadhams) [inline src/das-employer-favourites.ApplicationServices/Infrastructure/FatApiRepository.cs]: Casing

File: src/web/EmployerFavouritesWeb/nlog.config [hunk 15]
```diff
@@ -13,9 +13,9 @@
   <!-- the targets to write to -->
   <targets>
 
-    <target xsi:type="Redis" name="redis" connectionStringName="Redis" environmentKeyName="EnvironmentName" appName="das-employer-favourites-web" includeAllProperties="true" layout="${message}">
+    <target xsi:type="Redis" name="redis" connectionStringTitle="Redis" environmentKeyTitle="EnvironmentTitle" appTitle="das-employer-favourites-web" includeAllProperties="true" layout="${message}">
```
Comment (leewadhams) [inline src/web/EmployerFavouritesWeb/nlog.config]: Some weird renaming. Global rename error??

**Review by leewadhams — CHANGES_REQUESTED** [review]

File: src/api/EmployerFavouritesApi/nlog.config (hunk 12 above)
Comment (desderry) [reply src/api/EmployerFavouritesApi/nlog.config]: Resharper renaming....

**Review by leewadhams — APPROVED** [review]

**Pull request merged**